sage_code/type_two_primes.py

```diff
@@ -53,33 +59,58 @@
 GENERIC_UPPER_BOUND = 10**30
 
 
-def LLS(p):
-    return (log(p) + 9 + 2.5 * (log(log(p))) ** 2) ** 2
+def bound_on_largest_root(d, b):
+    """From Maarten's email"""
+    return RR(exp(-d*lambert_w(-1,-1/(d*b))))
 
 
-def get_type_2_uniform_bound(ecdb_type):
+def compute_S_d(d):
 
-    if ecdb_type == "LSS":
-        BOUND_TERM = (log(x) + 9 + 2.5 * (log(log(x))) ** 2) ** 2
-    elif ecdb_type == "BS":
-        # BOUND_TERM = (4*log(x) + 10)**2
-        # BOUND_TERM = (3.29*log(x) + 2.96 + 4.9)**2
-        BOUND_TERM = (1.881 * log(x) + 2 * 0.34 + 5.5) ** 2
-    else:
-        raise ValueError("argument must be LSS or BS")
+    g = 4 * log(x) + 10
 
-    f = BOUND_TERM**6 + BOUND_TERM**3 + 1 - x
+    f_BS = x - (g ** (4*d) + g ** (2*d) + 1)
 
+    largest_root_bound = max(bound_on_largest_root(4*d, 4),bound_on_largest_root(4*d, 5))
+
+    bound_BS = find_root(f_BS, 10, largest_root_bound)
+    
+    lls = (log(x) + 9 + 2.5 * (log(log(x))) ** 2) ** 2
     try:
-        bound = find_root(f, 1000, GENERIC_UPPER_BOUND)
-        return ceil(bound)
+        bound_const = find_root(10**9 - lls, 10**6, largest_root_bound)
     except RuntimeError:
-        warning_msg = (
-            "Warning: Type 2 bound for quadratic field with "
-            "discriminant {} failed. Returning generic upper bound"
-        ).format(5)
-        print(warning_msg)
-        return GENERIC_UPPER_BOUND
+        bound_const = largest_root_bound
```

Comment:
I don't fully get the logic here. First of all this `bound_const` doesn't depend on d.  So I think it would be better to just precompute and hard code `bound_constant` and add a comment saying "x>bound_constant ensures lls**2 > 10^9". This way we don't need the try except.

Or if you want to keep the computation here, just don't pass `largest_root_bound`  as a second parameter, but just some hardcoded constant so that we know we do not need the except.

Reply:
the logic in these two `try`/`except` blocks is: we want to find where `10**9` equals `lls**2` because we want to take the maximum in light of Remark 5.6; i call that crossover point `bound_const`, and then the true `bound_lls` arises from finding a root larger than `bound_const` and taking `max(bound_const, bound_LLS)`. I agree it's a bit convoluted, so will take your approach